core/node/proof_data_handler/src/tee_request_processor.rs

```diff
@@ -47,26 +47,62 @@ impl TeeRequestProcessor {
             .await
             .map_err(RequestProcessorError::Dal)?;
 
-        let l1_batch_number_result = connection
-            .tee_proof_generation_dal()
-            .get_next_batch_to_be_proven(request.tee_type, self.config.proof_generation_timeout())
-            .await
-            .map_err(RequestProcessorError::Dal)?;
-
-        let l1_batch_number = match l1_batch_number_result {
-            Some(number) => number,
-            None => return Ok(Json(TeeProofGenerationDataResponse(None))),
-        };
-
-        let tee_verifier_input: TeeVerifierInput = self
-            .blob_store
-            .get(l1_batch_number)
-            .await
-            .map_err(RequestProcessorError::ObjectStore)?;
+        loop {
+            let l1_batch_number = match connection
+                .tee_proof_generation_dal()
+                .get_next_batch_to_be_proven(
+                    request.tee_type,
+                    self.config.proof_generation_timeout(),
+                )
+                .await
+                .map_err(RequestProcessorError::Dal)?
+            {
+                Some(number) => number,
+                None => return Ok(Json(TeeProofGenerationDataResponse(None))),
+            };
+
+            match self.get_blob(l1_batch_number).await {
+                Ok(input) => {
+                    return Ok(Json(TeeProofGenerationDataResponse(Some(Box::new(input)))));
+                }
+                Err(ObjectStoreError::KeyNotFound(_)) => {
+                    tracing::warn!(
+                        "Blob for batch number {} has not been found in the object store. Marking the job as skipped.",
+                        l1_batch_number
+                    );
+                    connection
```

Comment:
> In general, skipped has different semantics in our codebase (skipped on purpose, not because input was missing).

I'm fine with renaming it however you like. We just need to make sure we can gracefully handle cases where the corresponding blob for the current batch is missing. We've already had an issue on stage where batches had no corresponding GCP blob, causing the TEE prover to panic.

> I'm not in the loop for this project, but why would TEE try to prove such old proofs? Where does it start? Does it make sense to waste compute on jobs that we know are not valuable at all?

We currently compute TEE proofs for all the batches in the database, starting from the very beginning. Maybe we should consider starting from a preconfigured batch instead of the oldest one.

Here’s how it works right now:
1. We have a [`tee-prover`](https://github.com/matter-labs/zksync-era/blob/main/core/bin/zksync_tee_prover/src/tee_prover.rs) standalone binary that runs an [infinite loop](https://github.com/matter-labs/zksync-era/blob/56d8ee8c0546cc26d412b95cb72bbb1b9a3a6580/core/bin/zksync_tee_prover/src/tee_prover.rs#L178-L218), pulling inputs for proving the next batches from the [`data_proof_handler`](https://github.com/matter-labs/zksync-era/blob/main/core/node/proof_data_handler/src/lib.rs).
1. The `data_proof_handler` [gives](https://github.com/matter-labs/zksync-era/blob/56d8ee8c0546cc26d412b95cb72bbb1b9a3a6580/core/lib/dal/src/tee_proof_generation_dal.rs#L21-L59) the inputs for the next batch to be proven.
1. The `tee-prover` [computes](https://github.com/matter-labs/zksync-era/blob/56d8ee8c0546cc26d412b95cb72bbb1b9a3a6580/core/bin/zksync_tee_prover/src/tee_prover.rs#L117) a TEE proof for the fetched inputs and [sends it back](https://github.com/matter-labs/zksync-era/blob/56d8ee8c0546cc26d412b95cb72bbb1b9a3a6580/core/bin/zksync_tee_prover/src/tee_prover.rs#L118-L125) to the proof handler to be saved in the database.

Reply:
> This sounds very sensible. Simply mark everything up to some point (manually, DB update) as skipped and then let the process flow by itself.

I’d really prefer to automate as much as possible. This PR adds automatic skipping. The other option would be manually running an `UPDATE` SQL query after each new DB re-deployment:
```
UPDATE tee_proof_generation_details SET status = 'skipped' WHERE l1_batch_number < 58300 ;
```
(this is the exact SQL query we ran manually when we realized that all batches `< 58300` don’t have matching blobs in GCP)

Whether we decide to manually skip batches below a certain number or not, we still need an easy way to handle missing blobs in GCP for any batch, old or new. This PR offers a way to do that.

Reply:
> due to some software bug or changes on retention policy -- we intend to drive that closer to 2 days, if your TEE is always 2 days behind, you'll skip all proofs

Good point too. 🤔

I don't have a strong opinion on how we should handle missing GCP blobs. This PR implements just one of many possible strategies to address this scenario. I'll wait for more review comments to see which approach is most preferable.

This PR isn't super urgent since we're currently working around the issue by manually running [the above SQL query](https://github.com/matter-labs/zksync-era/pull/2674#discussion_r1721896574). CC: @RomanBrodetski, @thomasknauth (and @popzxc, @haraldh when they return from vacation)